### ISSUE TITLE
Cancellation of promise collections

### ISSUE THREAD

**Issue by clue** [reactphp/promise#33]
What is the recommended way to cancel promise collections?

This library provides several promise collection primitives, which all have valid use cases.

One might assume the following is the way to go:

```php
$promises = array(
    accessSomeRemoteResource(),
    accessSomeRemoteResource(),
    accessSomeRemoteResource()
);

$promise = \React\Promise\all($promises);

// what does this do?
$promise->cancel();
```

How does this cope with cancellation support?

This is currently not implemented, i.e. calling `cancel()` currently has *no* effect.

Should we consider adding this?

Also, if yes, how could cancellation support work for the other primitives?

**Comment by Adirelle** [comment]
Bumped into this a few days ago.

In my opinion, the collection functions should accept an optional argument to define the cancellable behavior, with these possible values:
* `NEVER`: never cancels upstream promises; this is the default, legacy behavior.
* `ON_CANCEL`: try and cancel upstream promises when the downstream promise is cancelled.
* `ON_REJECT`: try and cancel upstream promises when the downstream promise is cancelled or rejected.
* `ALWAYS`: always try and cancel upstream promises.

Here are the effect of `ALWAYS` behavior with different functions when the downstream promise is fulfilled:
* `all`, `reduce`, `map`: does nothing,
* `race`, `some`, `any`: cancels all upstream promises that haven't been fulfilled or rejected.

I'm not sure how to handle `resolve()` when it generates an ExtendedPromiseInterface for an existing Promise.

**Comment by jsor** [comment]
Implemented in #36

**Comment by jsor** [comment]
Released in [v2.3.0](https://github.com/reactphp/promise/releases/tag/v2.3.0).
